### ISSUE TITLE
Streaming aggregation.

### ISSUE THREAD

**Issue by ryzhyk** [TimelyDataflow/differential-dataflow#296]
Given an input collection of integers, I'd like to output the sum of all values received at the current or earlier timestamp.  I would like to do this without storing the entire input set, by only keeping track of the previous value of the sum and adding the sum of new elements to it.  I may have figured out part of the puzzle, but was not able to implement it completely inside DD.

I use `differentiate` to capture and aggregate all new values for each timestamp:

```
// Input values.
let (value_input, values) = scope.new_collection();

scope.scoped::<AltNeu<usize>,_,_>("Sum", |inner| {
    let d_values = values.differentiate(inner);
    // Sum-up all values for each timestamp.
    let new_sum = d_values.map(|x| ((), x)).arrange_by_key().reduce(|_, vals, out| {
         let result: u32 = vals.iter().map(|v| v.0 * v.1).sum();
         out.push((result, 1));
    });

    // Not sure how to complete this.
    ...
}
```

Next, I want to compute the `full_sum` collection with a single update for each timestamp equal to the sum of all elements received so far, and to add the _previous_ value of this collection to `new_sum` to obtain the new total sum. I figured that I can use `delay` to implement the "_previous_ value" part, but I am struggling with the fact that the definition of `full_sum` is recursive.  In pseudocode:

```
let full_sum = new_sum.join_map(full_sum.delay(), |_, new, old| new + old).
```

However, this does not look like regular DD recursion, as there's no fix point computation here: the right-hand side of the equation contains the `full_sum` update from the previous timestamp, so the computation should complete in one shot.

I suppose I could output the value of `full_sum` at each timestamp and feed it back as input to DD.  This would work, but it feels like there must be a way to do this completely in DD without relying on the user to close the loop.

Thanks!

**Comment by frankmcsherry** [comment]
I may be confused, but isn't this just what `count` does? It takes whatever you've put in your `Diff` field, numbers perhaps, and accumulates them over time, reporting the accumulation for this and all prior times.

**Comment by ryzhyk** [comment]
Sorry, my example was too specific in the interests of clarity. I am looking for a general way to fold collections over time, i.e., replace sum with an arbitrary function that takes accumulator and a new element and computes the new accumulator value. This function may depend on other collections, e.g., I would like to be able to add an external signal to reset the counter. I realize this is not a typical dd scenario. It is partly inspired by your blog post on calculus operators and the as_of use case that breaks the main dd invariant that the output is a function of the input snapshot, but does so for good reasons :)

**Comment by frankmcsherry** [comment]
One proposal for this is to have a `Variable` feed back to the input to you operator, and have it introduce the corresponding removal of base data and introduction of aggregates. Let's say you wanted to implement `sum` but didn't know about the diff field and monoids and stuff. You could wire up a feedback that feeds back a retraction of some number of inputs and an insertion of their sum.

I think generally with an arbitrary state machine, you could have an input type that was an enum of a state and a symbol, whose `reduce` implementation was to play the symbols forward from the state (ideally the symbols have some integer to clarify in which order). At any point you could feed back the state update and the retraction of some symbols, which effectively acts as garbage collection.

Does that sound plausible?

**Comment by ryzhyk** [comment]
Thank you! I think I understand, let me give it a try.  What would you expect the memory footprint of this to be?

**Comment by ryzhyk** [comment]
> Thank you! I think I understand, let me give it a try.

I take it back, I don't :(

I wrote the following, which replaces the contents of the collection with the sum of its elements at each iteration.  It produces correct results, i.e., `sums` contains the sum of all values in the input stream after each timestamp, but I think it recomputes the sum from scratch at each iteration.  As the clock rolls forward, it takes longer to evaluate every subsequent timestamp.

```
            let (vals_input, vals): (_, Collection<_,_,isize>) = scope.new_collection();

            let sums = vals.iterate(|c| {
                c.map(|v| ((), v))
                 .reduce::<_,isize,isize>(|_, vs, out| {
                     // Compute the sum of all inputs.
                     let mut sum: isize = 0;
                     for (v, w) in vs {
                         sum += (**v) * w;
                     };
                     out.push((sum, 1));
                 })
                 .map(|(_, v)| v)
            });

            sums.inspect(|x| println!("Sums: {:?}", x))
                .probe_with(&mut probe);
```

**Comment by frankmcsherry** [comment]
Unless I misunderstand the harness, you are going to keep all records around for all iterations because DD is unable to compact iterative computations across iterations, until you close the input. If you add a new record to `vals_input`, the way DD will re-do the iterative computation is to walk back through each of the iterations.

I might have misunderstood your request. What role does iteration play in your goal?

**Comment by ryzhyk** [comment]
> I might have misunderstood your request. What role does iteration play in your goal?

It doesn't.  You mentioned `Variable` in your previous reply and I took it to mean iteration.  My bad. I see now that variables are used to implement iteration but can be used for other forms of feedback. Let me try again!

**Comment by frankmcsherry** [comment]
Yup, sorry! I meant a `Variable` in a non-iterative context.

**Comment by frankmcsherry** [comment]
If it helps, we have an example of doing this in Materialize, where we feed back retractions for TopK computation, once we can determine that a record will never be in the top k. It looks like
```rust
                use differential_dataflow::operators::iterate::Variable;
                let delay = std::time::Duration::from_nanos(10_000_000_000);
                let retractions = Variable::new(&mut ok_input.scope(), delay.as_millis() as u64);
                let thinned = ok_input.concat(&retractions.negate());
                let result = build_topk(thinned, group_key, order_key, *offset, *limit, arity);
                retractions.set(&ok_input.concat(&result.negate()));
```

**Comment by ryzhyk** [comment]
It works, thanks heaps! 

Here is my code for computing the sum:

```
            // Adjust `vals` by replacing old values with their sum.
            let adjustment = Variable::new(scope, 1);
            let adjusted = vals.concat(&adjustment);
            let sum = adjusted
                .map(|x| ((), x))
                .reduce::<_, isize, isize>(|_, vs, out| {
                     let mut sum: isize = 0;
                     for (v, w) in vs {
                         sum += (**v) * w;
                     };
                     out.push((sum, 1));
                })
                .map(|(_, v)| v);
            adjustment.set(&vals.negate().concat(&sum));
```

**Comment by ryzhyk** [comment]
A follow-up question: Using variables without recursion causes output updates with future timestamps. In DDlog we don't want to see those until we advance the epoch explicitly.  This is usually not a problem in practice, because we apply `consolidate` before `probe`, making sure that any output can only be produced once DD knows that no more changes are expected at the current timestamp, and since most output relations depend on at least one current input, outputs don't show up ahead of time.  In the exotic case when an output relation only depends on a variable, we can artificially delay the output by concatenating this relation with an always-empty input relation, or just forbid such relations altogether.

Is my reasoning above sound or am I making too strong assumptions about `consolidate`?

**Comment by frankmcsherry** [comment]
I think what you've said is correct. I would guess most uses of fed-back `Variable` data gets merged with other things, but if not you can merge with an empty stream and consolidate that, yup!

**Comment by ryzhyk** [comment]
Thanks again for helping out!  We already have streaming joins (#294) and now also adding stream aggregation as discussed here. Combined with "regular" DD relational capabilities, this is really powerful for stream analytics tasks.
